Added blog for announcing Istio client-go

nrjpoddar opened the pull request:
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[x] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

nrjpoddar commented:
@ayj @jasonwzm FYI

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | a903cd73faa2da3268ec968a5b5633038a22a0e8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1643) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+---
+title: "Announcing Istio client-go"
```
```suggestion
title: Announcing Istio client-go
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+---
+title: "Announcing Istio client-go"
+description: "Using Istio clien-go to gain programmatic access to Istio resources."
```
```suggestion
description: Getting programmatic access to Istio resources
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+target_release: 1.4
+---
+
+We are pleased to announce the official release of Istio
```
```suggestion
We are pleased to announce the initial release of the Istio
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+---
+
+We are pleased to announce the official release of Istio
+[client go](https://github.com/istio/client-go) repository which enables users
```
```suggestion
[client go](https://github.com/istio/client-go) repository which enables developers
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+to gain programmatic access to Istio APIs in a Kubernetes environment. The
+generated Kubernetes informers and client set in this repository makes it easy
+for users to create controllers and perform Create, Read, Update and Delete
+(CRUD) operations for all Istio Customer Resource Definitions (CRDs).
```
```suggestion
(CRUD) operations for all Istio Custor Resource Definitions (CRDs).
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+for users to create controllers and perform Create, Read, Update and Delete
+(CRUD) operations for all Istio Customer Resource Definitions (CRDs).
+
+This was a highly requested functionality by many Istio users, as is evident
```
```suggestion
This was a highly requested functionality by many Istio developers, as is evident
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+This was a highly requested functionality by many Istio users, as is evident
+from the feature requests on the clients generated by [Aspen Mesh
+Istio client go](https://github.com/aspenmesh/istio-client-go) and the [Knative
+project](https://github.com/knative/pkg/tree/a805b647f3f2aa6cec2f0490680377e5854775d9/apis/istio).
```
Can we use "master" here?

nrjpoddar replied:
We can, I know knative is going to remove this package soon, so I didn't want to knowingly add a link that will be removed soon.

I'm fine either way, your call.

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+from the feature requests on the clients generated by [Aspen Mesh
+Istio client go](https://github.com/aspenmesh/istio-client-go) and the [Knative
+project](https://github.com/knative/pkg/tree/a805b647f3f2aa6cec2f0490680377e5854775d9/apis/istio).
+If you're currently using one of these above mentioned clients, you can easily
```
```suggestion
If you're currently using one of the above mentioned clients, you can easily
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+switch to using [Istio client go](https://github.com/istio/client-go) like
+this:
+
+{{< text plain>}}
```
```suggestion
{{< text go>}}
```

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | 91241fa047c696aeae1324229b7f8458642ba5ca | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1646) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+}
+{{< /text >}}
+
+You can find a more in-depth example [here](https://github.com/istio/client-go/blob/7b65e62d85b04207a1b04ea047dfe635a1e31cb9/cmd/example/client.go).
```
```suggestion
You can find a more in-depth example [here](https://github.com/istio/client-go/blob/{{< source_branch_name >}}/cmd/example/client.go).
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+
+## Useful tools created for generating Istio client-go
+
+If you're one of the users who's wondering why did it take so long or why was it
```
```suggestion
If you're wondering why it took so long or why was it
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+
+If you're one of the users who's wondering why did it take so long or why was it
+difficult to generate this client set, this section is for you. In Istio, we use
+Proto specifications to write APIs which are then converted to Go definitions
```
```suggestion
[protobuf](https://developers.google.com/protocol-buffers) specifications to write APIs which are then converted to Go definitions
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+If you're one of the users who's wondering why did it take so long or why was it
+difficult to generate this client set, this section is for you. In Istio, we use
+Proto specifications to write APIs which are then converted to Go definitions
+using the Proto tool chain. There are three major challenges which you might
```
```suggestion
using the protobuf tool chain. There are three major challenges which you might
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+difficult to generate this client set, this section is for you. In Istio, we use
+Proto specifications to write APIs which are then converted to Go definitions
+using the Proto tool chain. There are three major challenges which you might
+face if you're trying to generate Kubernetes client set from a Proto generated API:
```
```suggestion
face if you're trying to generate Kubernetes client set from a protobuf-generated API:
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+* **Creating Kubernetes Wrapper Types** - Kubernetes [client generation](https://github.com/kubernetes/code-generator/tree/master/cmd/client-gen)
+library only works for Go objects which follow the Kubernetes object
+specification for e.g. [Authentication Policy Kubernetes
+Wrappers](https://github.com/istio/client-go/blob/7b65e62d85b04207a1b04ea047dfe635a1e31cb9/pkg/apis/authentication/v1alpha1/types.gen.go).
```
```suggestion
Wrappers](https://github.com/istio/client-go/blob/{{< source_branch_name >}}/pkg/apis/authentication/v1alpha1/types.gen.go).
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+which can automatically create the Kubernetes types based on annotations.
+The annotations which are parsed by this tool and various options
+are explained in the [README](https://github.com/istio/tools/blob/master/cmd/kubetype-gen/README.md).
+Note that if you're using Proto tools to generate Go types, you would need to
```
```suggestion
Note that if you're using protobuf tools to generate Go types, you would need to
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+The annotations which are parsed by this tool and various options
+are explained in the [README](https://github.com/istio/tools/blob/master/cmd/kubetype-gen/README.md).
+Note that if you're using Proto tools to generate Go types, you would need to
+add these annotations as comments in the Proto files, so that the comments are
```
```suggestion
add these annotations as comments in the proto files, so that the comments are
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+which is a `protoc` plugin and can automatically create `deepcopy` functions
+based on annotations using the Proto library utility [Proto
+Clone](https://godoc.org/github.com/golang/protobuf/proto#Clone). Here's an
+[example](https://github.com/istio/api/blob/e51134872853a42ff015106ed79d00115e983fba/authentication/v1alpha1/policy_deepcopy.gen.go)
```
```suggestion
[example](https://github.com/istio/api/blob/{{< source_branch_name >}}/authentication/v1alpha1/policy_deepcopy.gen.go)
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+of the generated `deepcopy` methods.
+
+* **Marshaling and Unmarshaling types to/from JSON** - For the types generated
+from Proto definitions, it is often problematic to use the default Golang JSON
```
```suggestion
from proto definitions, it is often problematic to use the default Go JSON
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+
+* **Marshaling and Unmarshaling types to/from JSON** - For the types generated
+from Proto definitions, it is often problematic to use the default Golang JSON
+encoder/decoder as there are various fields like Proto `oneof` which requires
```
```suggestion
encoder/decoder as there are various fields like protobuf's `oneof` which requires
```

geeknoid commented on content/en/blog/2019/announcing-istio-client-go/index.md:
```diff
@@ -0,0 +1,146 @@
+name might serialize/deserialize to different field names depending on the
+encoder/decoder as the Go struct tag are [generated
+differently](https://github.com/istio/istio/issues/17600).
+It is always recommended to use Proto primitives for
```
```suggestion
It is always recommended to use protobuf primitives for
```

geeknoid commented on .spelling:
```diff
@@ -223,6 +225,7 @@ GKE-Istio
 GKE-Workloads
 GlueCon
 Gmail
+Golang
```
No needed.

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | d1f7689e3d0eed67d617caa54baf4ef8023cd529 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1653) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | bafb2d379770b0d269fe5ab63775316413736f5b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1654) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | 0aa9ffdeadd49e18d667ea4409c409cc4e803f04 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1657) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | 7c9839a0fe1618522c715cfb4ad8a13c92de921d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1660) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nrjpoddar commented:
@geeknoid thank you so much for quick turnaround on this, I have addressed all comments except https://github.com/istio/istio.io/pull/5566/files#r344528735

istio-testing commented:
@nrjpoddar: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | 3f56ff38a0f1d562217b48339401f2cb75c8d852 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5566/lint_istio.io/1663) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
